cli_tools/gce_windows_upgrade/upgrader/workflows.go

```diff
@@ -0,0 +1,492 @@
+				NoCleanup: true,
+			},
+		},
+	}
+
+	stepDetachOldOSDisk, err := daisyutils.NewStep(w, "detach-old-os-disk", stepCreateNewOSDisk)
+	if err != nil {
+		return err
+	}
+	stepDetachOldOSDisk.DetachDisks = &daisy.DetachDisks{
+		&daisy.DetachDisk{
+			Instance:   u.InstanceURI,
+			DeviceName: daisyutils.GetDeviceURI(u.project, u.zone, u.osDiskDeviceName),
+		},
+	}
+
+	stepAttachNewOSDisk, err := daisyutils.NewStep(w, "attach-new-os-disk", stepDetachOldOSDisk)
+	if err != nil {
+		return err
+	}
+	stepAttachNewOSDisk.AttachDisks = &daisy.AttachDisks{
+		&daisy.AttachDisk{
+			Instance: u.InstanceURI,
+			AttachedDisk: compute.AttachedDisk{
+				Source:     u.newOSDiskName,
+				DeviceName: u.osDiskDeviceName,
+				AutoDelete: u.osDiskAutoDelete,
+				Boot:       true,
+			},
+		},
+	}
+
+	stepCreateInstallDisk, err := daisyutils.NewStep(w, "create-install-disk", stepAttachNewOSDisk)
+	if err != nil {
+		return err
+	}
+	stepCreateInstallDisk.CreateDisks = &daisy.CreateDisks{
+		&daisy.Disk{
+			Disk: compute.Disk{
+				Name:        u.installMediaDiskName,
+				Zone:        u.zone,
+				Type:        "pd-ssd",
+				SourceImage: "projects/compute-image-tools/global/images/family/windows-install-media",
+			},
+			Resource: daisy.Resource{
+				ExactName: true,
+				NoCleanup: true,
+			},
+		},
+	}
+
+	stepAttachInstallDisk, err := daisyutils.NewStep(w, "attach-install-disk", stepCreateInstallDisk)
+	if err != nil {
+		return err
+	}
+	stepAttachInstallDisk.AttachDisks = &daisy.AttachDisks{
+		&daisy.AttachDisk{
+			Instance: u.InstanceURI,
+			AttachedDisk: compute.AttachedDisk{
+				Source:     u.installMediaDiskName,
+				AutoDelete: true,
+			},
+		},
+	}
+	prevStep = stepAttachInstallDisk
+
+	// 'windows-startup-script-url' exists, backup it
+	if u.windowsStartupScriptURLBackup != nil {
+		fmt.Printf("\nDetected current '%v', value='%v'. Will backup to '%v'.\n\n", metadataKeyWindowsStartupScriptURL,
+			*u.windowsStartupScriptURLBackup, metadataKeyWindowsStartupScriptURLBackup)
+
+		stepBackupScript, err := daisyutils.NewStep(w, "backup-script", stepAttachInstallDisk)
+		if err != nil {
+			return err
+		}
+		stepBackupScript.UpdateInstancesMetadata = &daisy.UpdateInstancesMetadata{
+			&daisy.UpdateInstanceMetadata{
+				Instance: u.InstanceURI,
+				Metadata: map[string]string{metadataKeyWindowsStartupScriptURLBackup: *u.windowsStartupScriptURLBackup},
+			},
+		}
+		prevStep = stepBackupScript
+	} else {
+		if !u.windowsStartupScriptURLBackupExists {
+			fmt.Printf("\nNo existing startup script (metadata '%v') detected. Won't backup it.\n\n", metadataKeyWindowsStartupScriptURL)
+		}
+	}
+
+	stepSetScript, err := daisyutils.NewStep(w, "set-script", prevStep)
+	if err != nil {
+		return err
+	}
+	stepSetScript.UpdateInstancesMetadata = &daisy.UpdateInstancesMetadata{
+		&daisy.UpdateInstanceMetadata{
+			Instance: u.InstanceURI,
+			Metadata: map[string]string{metadataKeyWindowsStartupScriptURL: "${SOURCESPATH}/upgrade_script.ps1"},
```

Comment:
This is going to require GCS read permission on your upgrading instance. Do  you need that permission for other items? If not, you can put the script content inline as a metadata item.

Reply:
I don't want to use that work around since copying script from GCS is very common for our tools. If we claimed we don't need GCS read permission, however in the future when the logic became more complex and we have to copy multiple scripts, we will have to ask for an additional permission, which is a kind of regression and tough to rollout.